source/extensions/http/cache/file_system_http_cache/cache_file_fixed_block.cc

```diff
@@ -0,0 +1,94 @@
+#include "source/extensions/http/cache/file_system_http_cache/cache_file_fixed_block.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/common/safe_memcpy.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+namespace FileSystemHttpCache {
+
+namespace {
+// The expected first four bytes of the header - if fileId() doesn't match ExpectedFileId
+// then the file is not a cache file and should be removed from the cache.
+// Beginning of file should be "CACH".
+std::array<char, 4> ExpectedFileId = {'C', 'A', 'C', 'H'};
+
+// The expected next four bytes of the header - if cacheVersionId() doesn't match
+// ExpectedCacheVersionId then the file is from an incompatible cache version and should
+// be removed from the cache.
+// Next 4 bytes of file should be "0000".
+std::array<char, 4> ExpectedCacheVersionId = {'0', '0', '0', '0'};
+
+// Deserialize 4 bytes into a uint32_t.
+uint32_t deserializeUint32(const char* p) {
+  // chars must be cast to uint8_t first so that e.g. -1 becomes 0xff, before casting to uint32_t
+  // - otherwise -1 becomes 0xffffffff.
+  return static_cast<uint32_t>(static_cast<uint8_t>(p[0])) << 24 |
+         static_cast<uint32_t>(static_cast<uint8_t>(p[1])) << 16 |
+         static_cast<uint32_t>(static_cast<uint8_t>(p[2])) << 8 |
+         static_cast<uint32_t>(static_cast<uint8_t>(p[3]));
+}
+
+// Deserialize 8 bytes into a uint64_t.
+uint64_t deserializeUint64(const char* p) {
+  return static_cast<uint64_t>(deserializeUint32(p)) | deserializeUint32(p + 4);
+}
+
+// Serialize a uint32 into 4 bytes at pointer p.
+void serializeUint32(uint32_t value, char* p) {
+  p[0] = value >> 24;
+  p[1] = value >> 16 & 0xff;
+  p[2] = value >> 8 & 0xff;
+  p[3] = value & 0xff;
+}
+
+// Serialize a uint64 into 8 bytes at pointer p.
+void serializeUint64(uint64_t value, char* p) {
+  serializeUint32(value >> 32, p);
+  serializeUint32(value & 0xffffffff, p + 4);
+}
+
+} // namespace
+
+CacheFileFixedBlock::CacheFileFixedBlock() {
+  setFileId(ExpectedFileId);
+  setCacheVersionId(ExpectedCacheVersionId);
+  setHeadersSize(0);
+  setTrailersSize(0);
+  setBodySize(0);
+}
+
+void CacheFileFixedBlock::populateFromStringView(absl::string_view s) {
+  ASSERT(s.size() == size() && size() == 24);
+  // Serialize the values from the string_view s into the member values.
+  std::copy_n(s.begin(), 4, file_id_.begin());
```

Comment:
so much byte hacking. I really didn't mind the old code with the union and the calling of existing byte-order hacking methods; I just wanted it to be commented better.

Reply:
Counterpoint: really very little byte hacking at this point, and comparable to [other serialization code in the envoy codebase](https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/source/common/grpc/codec.cc?L19:43).

And significantly *less* dangerous than the union method that you originally didn't like, and no longer requires the half-safe memcpy function or any endian-ness conditions.

Reply:
We think there is too much byte hacking in Envoy and there have been various waves of work to eliminate them. Some of them need to stay when there is a strong need to do it for performance or whatever.

I apologize if it came across that I didn't like using unions -- I just wanted to clarify the rationale why you were doing it and suggested minor tweaks to make it more readable. I think I like it more than byte hacking like this :)

Reply:
It still ends up with unsafe byte-copying operations with the union option - there's at the very least one safeMemcpyUnsafeSrc involved to copy the data from a string_view to the union-object, even if we were to go with an inconsistent serialization option like you seem to prefer now.

There's really no getting away from the fact that serialization of integers into a fixed size data stream is going to involve byte-twiddling. It's just a matter of if you really think an explicitly unsafe copy operation and the awkwardness of a union is sufficiently better than moving 16 bytes with shifting. I don't think it is, I started changing it back to a union either as the class's internal representation *or* just as a temporary in the serialization functions, and IMO either way is less readable and looks less safe than this current implementation.

Arguably it could just be `ASSERT(incoming_string.size() == sizeof(*this)); safeMemcpyUnsafeSrc(incoming_string, *this)` with no inner union or struct, if we're not messing with bytes and not caring about inconsistent serialization, because within the same binary the serialization will be consistent and the class is acting as a struct itself, but to me that does not look safer, and it can definitely produce an inconsistent cache file layout that will at some point bite for debugging problems. And will go wrong in surprising ways if someone tries to add some other member to the class that isn't supposed to be part of the serialization, or make it virtual, etc.